File: backend/internal/blockparty/engine/game.go
package engine

// InitGame seeds a fresh board with two spawned tiles, mirroring the TS
// initGame: rngCallCount lands at 4 (two SpawnTile calls, 2 RNG draws each).
func InitGame(seed uint32, mod Modifier) State {
	empty := make(Board, 16)
	b1, rng1, cc1 := SpawnTile(empty, seed, 0, mod)
	b2, rng2, cc2 := SpawnTile(b1, rng1, cc1, mod)
	return State{
		Board:        b2,
		Score:        0,
		Rng:          rng2,
		RngCallCount: cc2,
		Modifier:     mod,
		Moves:        0,
		Over:         IsGameOver(b2),
	}
}

// Step applies a move and, only if the board changed, recomputes Over. A
// no-op move returns the input state unchanged.
func Step(s State, m Move) State {
	ns, changed := ApplyMove(s, m)
	if !changed {
		return s
	}
	ns.Over = IsGameOver(ns.Board)
	return ns
}

// Result is the terminal outcome of a Replay run.
type Result struct {
	Board        Board
	Score        int64
	RngCallCount int
	Over         bool
}

// Replay reproduces a full game from a seed and a sequence of moves.
func Replay(seed uint32, mod Modifier, moves []Move) Result {
	s := InitGame(seed, mod)
	for _, m := range moves {
		s = Step(s, m)
	}
	return Result{Board: s.Board, Score: s.Score, RngCallCount: s.RngCallCount, Over: s.Over}
}

File: frontend/src/game/engine/game.ts
import { spawnTile } from "./board";
import { applyMove } from "./move";
import { isGameOver } from "./gameover";
import type { Board, GameState, Modifier, Move } from "./types";

export function initGame(seed: number, modifier: Modifier): GameState {
  const empty: Board = new Array(16).fill(0);
  const first = spawnTile(empty, seed >>> 0, 0, modifier);
  const second = spawnTile(first.board, first.rng, first.rngCallCount, modifier);
  return {
    board: second.board,
    score: 0,
    rng: second.rng,
    rngCallCount: second.rngCallCount,
    modifier,
    moves: 0,
    over: isGameOver(second.board),
  };
}

export function step(state: GameState, move: Move): GameState {
  const next = applyMove(state, move);
  if (next === state) return state; // no-op passthrough
  return { ...next, over: isGameOver(next.board) };
}

export function replay(
  seed: number,
  modifier: Modifier,
  moves: Move[]
): { board: Board; score: number; rngCallCount: number; over: boolean } {
  let s = initGame(seed, modifier);
  for (const m of moves) s = step(s, m);
  return { board: s.board, score: s.score, rngCallCount: s.rngCallCount, over: s.over };
}
